pkg/core/object/metadata.go

```diff
@@ -197,3 +200,641 @@ func VerifyHeaderForMetadata(hdr object.Object) error {
+	errWrongKeyValDelim      = errors.New("wrong key-value delimiter")
+
+	// AttributeDelimiter is attribute key and value separator used in metadata DB.
+	AttributeDelimiter = []byte{0x00}
+	metaOIDPrefix      = []byte{metaPrefixID}
+
+	maxUint256 = new(big.Int).SetBytes([]byte{255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
+		255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255})
+	maxUint256Neg = new(big.Int).Neg(maxUint256)
+)
+
+// SearchCursor is a cursor used for continuous search in the DB.
+type SearchCursor struct {
+	// PrimarySeekKey defines the last key that was already handled.
+	PrimarySeekKey []byte
+	// PrimaryKeysPrefix defines what part of primary key is prefix.
+	PrimaryKeysPrefix []byte
+}
+
+// ParsedIntFilter is returned by [PreprocessIntFilters] to pass into the
+// [DB.Search].
+type ParsedIntFilter struct {
+	// AutoMatch means every existing value if acceptable for filters.
+	AutoMatch bool
+	// IntValue is parsed integer value from attribute.
+	IntValue *big.Int
+	// RawValue is raw attribute value in the original form.
+	RawValue []byte
+}
+
+// AttributeGetter provides access to attributes based on data base's indexes.
+type AttributeGetter interface {
+	// Get must return attribute value for a pair of object ID and attribute
+	// value. (nil, nil) must be returned if value is missing. err should mean
+	// only data base errors.
+	Get(oID []byte, attributeKey string) (attributeValue []byte, err error)
+}
+
+// AdditionalObjectChecker is an additional check that may be provided to
+// [MetaDataKVHandler]. `false` return value stops object from being recorded
+// to [SearchResult], `true` keeps searching without additional filtering.
+type AdditionalObjectChecker func(oid.ID) (match bool)
+
+// SearchResult is a [MetaDataKVHandler] operation's results holder.
+type SearchResult struct {
+	Objects             []client.SearchResultItem
+	UpdatedSearchCursor []byte
+	Err                 error
+}
+
+// MetaDataKVHandler returns a handler that filters out and writes search results
+// for Search NeoFS operation (see [client.SearchObjects] for details) based on
+// the provided arguments. resHolder must not be nil. additionalCheck may or may
+// not be nil, it will be called once on every object that matches all the previous
+// (filters-based) checks. fInt must have preparsed (from string attribute form)
+// integer values from the fs.
+// Returned handler must be called on a sorted list of key-value pairs without
+// any pair omission and/or duplication. Pairs must be common NeoFS indexes for
+// object header's fields. See [meta] (storage engine's package) for details.
+func MetaDataKVHandler(resHolder *SearchResult, attrGetter AttributeGetter, additionalCheck AdditionalObjectChecker, fs object.SearchFilters, fInt map[int]ParsedIntFilter, attrs []string, cursor *SearchCursor, count uint16) func(k, v []byte) bool {
```

Comment:
Scary as hell.

Reply:
what exactly?

Reply:
SearchV2, first of all. We're risking a lot here, I can't accept it to 0.45.2. The interface itself is quite complex.

Reply:
> The interface itself is quite complex.

will `iter.Seq` do the trick for you? (thread below)

Reply:
Not yet known. `iter.Seq` will have some other problems (DB transaction management).